core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java

```diff
@@ -487,4 +322,190 @@ public boolean isFinished()
+            collectMinMax = false;
+            if (state == ChannelState.MIN_MAX) {
+                state = ChannelState.NONE;
+            }
+            // Drop references to collected values.
+            minValues = null;
+            maxValues = null;
+        }
+
+        private Domain convertToDomain()
+        {
+            Block block = blockBuilder.build();
+            ImmutableList.Builder<Object> values = ImmutableList.builder();
+            for (int position = 0; position < block.getPositionCount(); ++position) {
+                Object value = readNativeValue(type, block, position);
+                if (value != null) {
+                    // join doesn't match rows with NaN values.
+                    if (!isFloatingPointNaN(type, value)) {
+                        values.add(value);
+                    }
+                }
+            }
+
+            // Inner and right join doesn't match rows with null key column values.
+            return Domain.create(ValueSet.copyOf(type, values.build()), false);
+        }
+
+        private void updateMinMaxValues(Block block, BlockPositionComparison comparison)
+        {
+            int minValuePosition = -1;
+            int maxValuePosition = -1;
+
+            for (int position = 0; position < block.getPositionCount(); ++position) {
+                if (block.isNull(position)) {
+                    continue;
+                }
+                if (minValuePosition == -1) {
+                    // First non-null value
+                    minValuePosition = position;
+                    maxValuePosition = position;
+                    continue;
+                }
+                if (comparison.compare(block, position, block, minValuePosition) < 0) {
+                    minValuePosition = position;
+                }
+                else if (comparison.compare(block, position, block, maxValuePosition) > 0) {
+                    maxValuePosition = position;
+                }
+            }
+
+            if (minValuePosition == -1) {
+                // all block values are nulls
+                return;
+            }
+            if (minValues == null) {
+                // First Page with non-null value for this block
+                minValues = block.getSingleValueBlock(minValuePosition);
+                maxValues = block.getSingleValueBlock(maxValuePosition);
+                return;
+            }
+            // Compare with min/max values from previous Pages
+            Block currentMin = minValues;
+            Block currentMax = maxValues;
+
+            if (comparison.compare(block, minValuePosition, currentMin, 0) < 0) {
+                minValues = block.getSingleValueBlock(minValuePosition);
+            }
+            if (comparison.compare(block, maxValuePosition, currentMax, 0) > 0) {
+                maxValues = block.getSingleValueBlock(maxValuePosition);
+            }
+        }
+    }
+
+    private void finishDomainCollectionIfNecessary()
```

Comment:
Can we move this method up to where the rest of this class methods are now that we don't reference it directly in ChannelFilter ?